frameless-runtime/src/utxo.rs

```diff
@@ -0,0 +1,253 @@
+///     MoneyUTXO // Configuration for MoneyUTXO
+///     KittiesUTXO // Configuration for KittiesUTXO
+///     ExistenceUTXO // Configuration for ExistenceUTXO
+/// )
+///
+
+// TODO: Configurable maybe when configuring overall UTXO Runtime?
+// For now hardcoded
+pub type OutputRef = H256;
+pub type Address = H256;
+pub type Value = Vec<u8>;
+pub type Sig = Vec<u8>;
+pub type TypeId = [u8; 4];
+pub type Redeemer = sp_core::H256;
+
+// pub type DispatchResult = Result<(), sp_runtime::DispatchError>;
+// Temporary should probably move to something like this above ^^
+pub type DispatchResult = Result<(), ()>;
+
+/// A single input references the output to be consumed or peeked at and provides some witness data, possibly a signature.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, parity_util_mem::MallocSizeOf))]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Default, Clone, Encode, Decode, Hash, Debug, TypeInfo)]
+pub struct Input {
+    /// A previously created output that will be consumed by the transaction containing this input.
+    pub output: OutputRef,
+    /// A witness proving that the output can be consumed by this input. In many cases including that of a basic cryptocurrency, this will be a digital signature.
+    pub witness: Sig,
+}
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, parity_util_mem::MallocSizeOf))]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Default, Clone, Encode, Decode, Hash, Debug, TypeInfo)]
+pub struct Output {
+    /// The address that owns this output. Based on either a public key or a Tuxedo Piece
+    pub redeemer: Redeemer,
+    /// The data associated with this output. In the simplest case, this will be a token balance, but could be arbitrarily rich state.
+    pub data: Value,
+    /// An Id for this type Such that we know how to encode or decode the 'data' field
+    pub data_id: TypeId,
+}
+
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, parity_util_mem::MallocSizeOf))]
+#[derive(PartialEq, Eq, PartialOrd, Ord, Default, Clone, Encode, Decode, Hash, Debug, TypeInfo)]
+pub struct Transaction {
+    /// The inputs refer to currently existing unspent outputs that will be consumed by this transaction
+    pub inputs: Vec<Input>,
+    /// Similar to inputs, Peeks refer to currently existing utxos, but they will be read only, and not consumed
+    pub peeks: Option<Vec<Input>>,
+    /// The new outputs to be created by this transaction.
+    pub outputs: Vec<Output>,
+}
+
+pub type Utxo = Output;
+pub type UtxoRef = OutputRef;
+
+pub trait Redeem {
+    fn redeem(self, tx: &[u8], witness: &[u8]) -> bool;
+}
+
+impl Redeem for Redeemer {
+    fn redeem(self, tx: &[u8], witness: &[u8]) -> bool {
+        let signature = match Signature::try_from(&witness[..]) {
+            Ok(sig) => sig,
+            Err(_) => return false,
+        };
+        sp_io::crypto::sr25519_verify(&signature, &tx, &Public::from_h256(self))
+    }
+}
+
+// TODO: Implement this for Each Tuxedo Piece
```

Comment:
We should think through the advantages to having a separate utxo set for each piece as opposed to a single utxo set. I'm not currently seeing advantages to having more than one, but I could just be missing them.

Reply:
I initially felt this way but I thought perhaps the nullify part might have different implementations depending on the Piece. I couldnt think of any at the time but just because I couldnt think of any I didnt want to assume it wasnt possible.